### ISSUE TITLE
Is the commit loss ignored?

### ISSUE THREAD

**Issue by githubnemo** [nakosung/VQ-VAE#2]
The current state is

```python

            loss = loss_rec + loss_e1 + self.vq_beta * loss_e2
            self.reset_grad()

            # For decoder
            loss.backward(retain_graph=True)

            # For encoder
            self.G.bwd()
```

where first the gradients for the decoder are computed (`loss.backward`) and then the gradients for the 'encoder' are computed (`self.G.bwd`). However, the `loss` term also includes a term that *should* be propagated through the encoder, namely the commitment term `loss_e2`. But wouldn't `self.G.bwd()` overwrite all gradients on the encoder?

**Comment by nakosung** [comment]
As you can see, there is no zero_grad between two backwards. By not zeroing grads, we could accumulate two gradients, so `self.G.bwd()` doesn't overwrite previous gradients.

https://discuss.pytorch.org/t/how-to-implement-accumulated-gradient-in-pytorch-i-e-iter-size-in-caffe-prototxt/2522

**Comment by ottonemo** [comment]
Ah, yes you are right. Should work then!

**Comment by ottonemo** [comment]
As a side note, your storage hook is almost equivalent to `term.retain_grad()` which also installs a backward hook and stores the gradient in `term.grad`.
